File: greektrans.js
// Greek character conversion utility

/// initialize table

import { table } from "./romanize.js";

function fromReversedTable(tableRev) {
    const table = {};
    for (const [key, value] of Object.entries(tableRev)) {
        for (const letter of value) {
            table[letter] = key;
        }
    }
    return table;
}

function capitalize(text) {
    return text[0].toUpperCase() + text.slice(1);
}

const romanizationTable = fromReversedTable(table.romanizationTableRev);
for (const ch of table.greekCapitalLetters) {
    romanizationTable[ch] = capitalize(romanizationTable[ch.toLowerCase()]);
}

const attributeCodes = {};
for (const [key, value] of Object.entries(table["attributeCodes"])) {
    attributeCodes[key] = String.fromCharCode(parseInt(value, 16));
}

const greekAttrsChar = {};
for (const [key, value] of Object.entries(attributeCodes)) {
    if (["ACUTE_ACCENT", "GRAVE_ACCENT", "GREEK_PERISPOMENI"].includes(key)) {
        greekAttrsChar[value] = attributeCodes.ACUTE_ACCENT;
    } else if (key == "DIAERESIS") {
        greekAttrsChar[value] = attributeCodes.DIAERESIS;
    } else {
        greekAttrsChar[value] = "";
    }
}

function isDecomposed(text) {
    return text == text.normalize("NFD");
}

function hasAttr(text, attr) {
    return text.normalize("NFD").includes(attributeCodes[attr]);
}

const combinationTable = {}
const diphthongs = {}
for (const [key, value] of Object.entries(table.combination)) {
    combinationTable[key] = value;
    if (isDecomposed(key)) {
        combinationTable[key.toUpperCase()] = value.toUpperCase();
    } else if (value[0] === "h" || hasAttr(key, "COMMA_ABOVE")) {
        combinationTable[capitalize(key)] = capitalize(value);
    }
    if (key.normalize("NFD").length == 2 && table.greekVowels.includes(key[0])) {
        diphthongs[key] = "α"
    }
}

function addUpper(table) {
    for (const [key, value] of Object.entries(table)) {
        table[key.toUpperCase()] = value.toUpperCase();
    }
}

addUpper(table.caron);
addUpper(table.dotMacron);

/// utility functions

export function isLetter(letter) {
    return table.greekLetters.includes(letter);
}

export function isVowel(letter) {
    return table.greekVowels.includes(letter);
}

export function isConsonant(letter) {
    return table.greekConsonants.includes(letter);
}

export function strip(text) {
    text = text.normalize("NFD");
    for (const ch in greekAttrsChar) {
        text = text.replaceAll(ch, "")
    }
    return text.normalize("NFC");
}

function replaces(table, text) {
    for (const [key, value] of Object.entries(table)) {
        if (key != value) text = text.replaceAll(key, value);
    }
    return text;
}

function* tokenize(text) {
    text = text.normalize("NFC");
    let token = "";
    let type = 0;
    for (const ch of text) {
        const t = isLetter(ch) ? 1 : 2;
        if (type !== t) {
            if (token) {
                yield [type, token];
                token = "";
            }
            type = t;
        }
        token += ch;
    }
    if (token) yield [type, token];
}

function prepareMonotonize(word) {
    const w = strip(word);
    if (Array.from(replaces(diphthongs, w)).filter(isVowel).length < 2) {
        return w;
    }
    return word;
}

export function monotonize(text) {
    let ret = "";
    for (const [type, token] of tokenize(text)) {
        if (type === 1) {
            ret += prepareMonotonize(token);
        } else {
            ret += token;
        }
    }
    ret = ret.normalize("NFD");
    ret = replaces(greekAttrsChar, ret);
    return ret.normalize("NFC");
}

function prepareRomanize(word) {
    let ret = "";
    if (word.length >= 1 && isVowel(word[0])) {
        if (word.length >= 2 && combinationTable[word.slice(0, 2)]) {
            ret = word.slice(0, 2);
            word = word.slice(2);
        } else {
            ret = word.slice(0, 1);
            word = word.slice(1);
        }
    }
    // check coronis: ex. κἀγώ = καὶ ἐγώ
    const psili = table.attributes.COMMA_ABOVE;
    for (const ch of word) {
        ret += ch;
        if (psili.includes(ch)) ret += "'";
    }
    return ret;
}

export function romanize(text, caron = true, dotMacron = true) {
    let ret = "";
    for (const [type, token] of tokenize(text)) {
        if (type === 1) {
            ret += prepareRomanize(token);
        } else {
            ret += token;
        }
    }
    ret = replaces(combinationTable, ret);
    ret = [...ret].map(letter => romanizationTable[letter] || letter).join("");
    if (caron) {
        ret = replaces(table.caron, ret);
    }
    if (dotMacron) {
        ret = replaces(table.dotMacron, ret);
    }
    return ret;
}

if (import.meta.main) {
    let file, monotonic = false;
    if (Deno && Deno.args && Deno.args.length) {
        if (Deno.args[0] == "-m") {
            monotonic = true;
            file = Deno.args[1];
        } else {
            file = Deno.args[0];
        }
    }
    if (!file) {
        console.error("usage: deno run --allow-read el-utils.js [-m] input_file");
        if (Deno) Deno.exit(1);
    } else {
        try {
            const text = Deno.readTextFileSync(file);
            console.log((monotonic ? monotonize : romanize)(text).trimEnd());
        } catch (e) {
            console.error(e);
            Deno.exit(1);
        }
    }
}

File: greektrans.py
# Greek character conversion utility

import unicodedata

### initialize table

from romanize import table

def from_reversed_table(table_rev):
    table = {}
    for key, value in table_rev.items():
        for letter in value:
            table[letter] = key
    return table

romanization_table = from_reversed_table(table["romanizationTableRev"])
for ch in table["greekCapitalLetters"]:
    romanization_table[ch] = romanization_table[ch.lower()].capitalize()

attribute_codes = {key: chr(int(value, 16)) for key, value in table["attributeCodes"].items()}
greek_attrs_char = {}
for key, value in attribute_codes.items():
    if key in ["ACUTE_ACCENT", "GREEK_PERISPOMENI"]:
        greek_attrs_char[value] = attribute_codes["ACUTE_ACCENT"]
    elif key == "DIAERESIS":
        greek_attrs_char[value] = attribute_codes["DIAERESIS"]
    else:
        greek_attrs_char[value] = ""

def is_decomposed(text):
    return text == unicodedata.normalize("NFD", text)

def has_attr(text, attr):
    return attribute_codes[attr] in unicodedata.normalize("NFD", text)

combination_table = {}
for key, value in table["combination"].items():
    combination_table[key] = value
    if is_decomposed(key):
        combination_table[key.upper()] = value.upper()
    elif value[0] == "h" or has_attr(key, "COMMA_ABOVE"):
        combination_table[key.capitalize()] = value.capitalize()

def add_upper(table):
    table.update({key.upper(): value.upper() for key, value in table.items()})

add_upper(table["caron"])
add_upper(table["dotMacron"])

### utility functions

def is_letter(letter):
    return letter in table["greekLetters"]

def is_vowel(letter):
    return letter in table["greekVowels"]

def is_consonant(letter):
    return letter in table["greekConsonants"]

def strip(text):
    text = unicodedata.normalize("NFD", text)
    for ch in greek_attrs_char:
        text = text.replace(ch, "")
    return unicodedata.normalize("NFC", text)

def replaces(table, text):
    for key, value in table.items():
        if key != value:
            text = text.replace(key, value)
    return text

def monotonize(text):
    text = unicodedata.normalize("NFD", text)
    text = replaces(greek_attrs_char, text)
    return unicodedata.normalize("NFC", text)

def prepare_romanize(word):
    ret = ""
    if len(word) >= 1 and is_vowel(word[0]):
        if len(word) >= 2 and word[:2] in combination_table:
            ret = word[:2]
            word = word[2:]
        else:
            ret = word[:1]
            word = word[1:]
    # check coronis: ex. κἀγώ = καὶ ἐγώ
    psili = table["attributes"]["COMMA_ABOVE"]
    for ch in word:
        ret += ch
        if ch in psili:
            ret += "'"
    return ret

def tokenize(text):
    text = unicodedata.normalize("NFC", text)
    token = ""
    type = 0
    for ch in text:
        t = 1 if is_letter(ch) else 2
        if type != t:
            if token:
                yield type, token
                token = ""
            type = t
        token += ch
    if token:
        yield type, token

def romanize(text, caron=True, dot_macron=True):
    ret = ""
    for type, token in tokenize(text):
        if type == 1:
            ret += prepare_romanize(token)
        else:
            ret += token
    ret = replaces(combination_table, ret)
    ret = "".join(map(lambda letter: romanization_table.get(letter, letter), ret))
    if caron:
        ret = replaces(table["caron"], ret)
    if dot_macron:
        ret = replaces(table["dotMacron"], ret)
    return ret

if __name__ == "__main__":
    import sys
    file, monotonic = None, False
    if len(sys.argv) > 1:
        if sys.argv[1] == "-m":
            monotonic = True
            file = sys.argv[2]
        else:
            file = sys.argv[1]
    if not file:
        print("usage: python el-utils.py [-m] input_file", file=sys.stderr)
        sys.exit(1)
    try:
        with open(file, "r", encoding="utf-8") as f:
            text = f.read()
        print((monotonize if monotonic else romanize)(text).rstrip())
    except Exception as e:
        print(e)
        sys.exit(1)
